src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -235,6 +250,93 @@ internal static void SetProperty(object resource, string propertyName, object va
             }
         }
 
+        internal static void SetInstanceAnnotations(ODataResource oDataResource, IODataInstanceAnnotationContainer instanceAnnotationContainer, 
+            ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
+        {
+            if(oDataResource.InstanceAnnotations != null)
+            {
+                foreach (ODataInstanceAnnotation annotation in oDataResource.InstanceAnnotations)
+                {
+                    AddInstanceAnnotationToContainer(instanceAnnotationContainer, deserializerProvider, readContext, annotation,string.Empty);
+                }
+            }
+
+            foreach(ODataProperty property in oDataResource.Properties)
+            {
+                if(property.InstanceAnnotations != null)
+                {
+                    foreach (ODataInstanceAnnotation annotation in property.InstanceAnnotations)
+                    {
+                        AddInstanceAnnotationToContainer(instanceAnnotationContainer, deserializerProvider, readContext, annotation, property.Name);
+                    }
+                }
+            }
+        }
+
+        private static void AddInstanceAnnotationToContainer(IODataInstanceAnnotationContainer instanceAnnotationContainer, ODataDeserializerProvider deserializerProvider, 
+            ODataDeserializerContext readContext, ODataInstanceAnnotation annotation, string propertyName)
+        {
+            EdmTypeKind type;
+            IEdmTypeReference propertyType = null;
+
+            object annotationValue = ConvertValue(annotation.Value, ref propertyType, deserializerProvider, readContext, out type);
+
+            if (string.IsNullOrEmpty(propertyName))
+            {
+                instanceAnnotationContainer.AddResourceAnnotation(annotation.Name, annotationValue);
+            }
+            else
+            {
+                instanceAnnotationContainer.AddPropertyAnnotation(propertyName,annotation.Name, annotationValue);
+            }
+        }
+
+        public static IODataInstanceAnnotationContainer GetAnnotationContainer(PropertyInfo propertyInfo, object resource)
+        {
+            IDelta delta = resource as IDelta;
+            object value;
+            if (delta != null)
+            {
+                delta.TryGetPropertyValue(propertyInfo.Name, out value);
+            }
+            else
+            {
+                value = propertyInfo.GetValue(resource);
+            }
+
+            IODataInstanceAnnotationContainer instanceAnnotationContainer = value as IODataInstanceAnnotationContainer;
+
+            if (instanceAnnotationContainer == null)
+            {
+                try
+                {
+                    if (propertyInfo.PropertyType == typeof(ODataInstanceAnnotationContainer) || propertyInfo.PropertyType == typeof(IODataInstanceAnnotationContainer))
```

Comment:
>(propertyInfo.PropertyType == typeof(ODataInstanceAnnotationContainer) || propertyInfo.PropertyType == type [](start = 23, length = 107)

If a new developer creates his own Container, so, this code can't process his type? 

You can use IsAssignFrom(...)

Reply:
This is for default or when they havent instantiated instance, the else part will handle if they creates their own container

Reply:
Right; this is a perf optimization to not use reflection to createinstance in the common case where we can create an instance of our own type.

---
In reply to: [489040284](https://github.com/OData/WebApi/pull/2219#discussion_r489040284) [](ancestors = 489040284)

Reply:
Actually, we are using this to decide if we can set the null property to an instance of our default InstanceAnnotationcontainer. We already know it has to implement IODataInstanceAnnotationContainer, we want to make sure that the property is not defined as a custom implementation of IODataInstanceAnnotationContainer.

---
In reply to: [489002262](https://github.com/OData/WebApi/pull/2219#discussion_r489002262) [](ancestors = 489002262)